src/drivers/unified_gamepad/driver.rs

```diff
@@ -0,0 +1,154 @@
+        let device = api.open_path(&c_path)?;
+        let info = device.get_device_info()?;
+        let vid = info.vendor_id();
+        let pid = info.product_id();
+        if vid != UNIFIED_CONTROLLER_VID || pid != UNIFIED_CONTROLLER_PID {
+            return Err(format!("Device '{path}' is not a Unified Controller: {vid}:{pid}").into());
+        }
+
+        Ok(Self {
+            device,
+            capabilities: None,
+            state: None,
+        })
+    }
+
+    /// Poll the device and read input reports
+    #[allow(dead_code)]
+    pub fn poll(&mut self) -> Result<Vec<Event>, Box<dyn Error + Send + Sync>> {
+        // Check the device for its capabilities
+        if self.capabilities.is_none() {
+            let mut buf = [0; INPUT_CAPABILITY_REPORT_SIZE];
+            buf[0] = ReportType::InputCapabilityReport as u8;
+            let _bytes_read = self.device.get_feature_report(&mut buf)?;
+            let report = InputCapabilityReport::unpack(&buf)?;
+            self.capabilities = Some(report);
+        }
+
+        // Read data from the device into a buffer
+        let mut buf = [0; INPUT_DATA_REPORT_SIZE];
+        let bytes_read = self.device.read(&mut buf[..])?;
+        let slice = &buf[..bytes_read];
+
+        // Handle the incoming input report
+        let events = self.handle_report(slice, bytes_read)?;
+
+        Ok(events)
+    }
+
+    #[allow(dead_code)]
+    fn handle_report(
+        &mut self,
+        buf: &[u8],
+        _bytes_read: usize,
+    ) -> Result<Vec<Event>, Box<dyn Error + Send + Sync>> {
+        // The first and second bytes contain the specification version
+        let major_ver = buf[0];
+        let minor_ver = buf[1];
+
+        // The major version indicates there are breaking changes.
+        if major_ver != UNIFIED_SPEC_VERSION_MAJOR {
+            return Err(format!("Device major version (v{major_ver}) is not compatible with this implementation (v{UNIFIED_SPEC_VERSION_MAJOR})").into());
+        }
+        // Minor versions are backwards compatible
+        if minor_ver > UNIFIED_SPEC_VERSION_MINOR {
+            return Err(format!("Device minor version (v{minor_ver}) is newer than this implementation supports (v{UNIFIED_SPEC_VERSION_MINOR})").into());
+        }
+
+        let report_type = ReportType::from(buf[2]);
+        match report_type {
+            ReportType::Unknown => (),
+            ReportType::InputCapabilityReport => {
+                let report = InputCapabilityReport::unpack(buf)?;
+                self.capabilities = Some(report);
+                // If the capabilities change, zero out the old state
+                self.state = None;
+            }
+            ReportType::InputDataReport => {
+                return self.handle_input_report(buf);
+            }
+            ReportType::OutputCapabilityReport => (),
+            ReportType::OutputDataReport => (),
+        }
+
+        Ok(vec![])
+    }
+
+    fn handle_input_report(
```

Comment:
I like this a lot. I think we can use a similar pattern in other drivers to simplify them.